src/dftbp/timedep/transcharges.F90

```diff
@@ -113,18 +135,156 @@ subroutine TTransCharges_init(this, iAtomStart, sTimesGrndEigVecs, grndEigVecs,
     !> index array for single particle excitations that are included
     integer, intent(in) :: win(:)
 
-    integer :: ia, ij, ii, jj, kk, ab, aa, bb
+    integer :: ia, ii, jj, ij, kk, ab, aa, bb, ss
     logical :: updwn
 
-    this%nTransitions = nTrans
-    this%nAtom = size(iAtomStart) - 1
+  #:if WITH_SCALAPACK
+    
+    integer :: iSpin, iAtom, nSpin, nOrbs, nLocRow, nLocCol
+    integer :: desc(DLEN_), iLoc, mLoc, iGlb, mGlb, jLoc, jGlb
+    real(dp), allocatable :: maskMat(:,:), workLocal(:,:), workGlobal(:,:,:)
+
+  #:endif
+    
+    this%nTransitions = nTrans  
+    this%nAtom = size(denseDesc%iAtomStart) - 1
     this%nMatUp = nMatUp
     this%nMatUpOccOcc = nXooUD(1)
     this%nMatUpVirVir = nXvvUD(1)
 
+  #:if WITH_SCALAPACK
+
+    nSpin = size(grndEigVecs, dim=3)
+     
+    if (tStoreSame) then
+
+      @:ASSERT(tStoreOccVir)
+      if (tFirstCall) then
+        @:ASSERT(.not.allocated(this%qCacheOccOcc))
+        allocate(this%qCacheOccOcc(this%nAtom, sum(nXooUD)))
+        @:ASSERT(.not.allocated(this%qCacheVirVir))
+        allocate(this%qCacheVirVir(this%nAtom, sum(nXvvUD)))
+      end if
+      
+    else
+      
+      this%tCacheChargesSame = .false.
+      
+    end if
+    
     if (tStoreOccVir) then
+      
+      if (tFirstCall) then
+        @:ASSERT(.not.allocated(this%qCacheOccVir))
+      else
+        ! A simple deallocate gives me an error, why is this%qCacheOccVir
+        ! deallocated if routine is re-entered?
+        if(allocated(this%qCacheOccVir)) deallocate(this%qCacheOccVir)
+      end if
 
-      @:ASSERT(.not.allocated(this%qCacheOccVir))
+      allocate(this%qCacheOccVir(this%nAtom, nTrans))
+      desc(:) = denseDesc%blacsOrbSqr
+      allocate(workGlobal(nOrb,nOrb,nSpin))
+      nLocRow = size(grndEigVecs,dim=1)
+      nLocCol = size(grndEigVecs,dim=2)
+      allocate(workLocal(nLocRow,nLocCol))
+      allocate(maskMat(nLocRow,nLocCol))
+
+      do iSpin = 1, nSpin
+        do iAtom = 1, this%nAtom
+          workGlobal = 0.0_dp
+          workLocal = 0.0_dp
+          maskMat = 0.0_dp
+          
+          do mLoc = 1, size(grndEigVecs, dim=1)
+            mGlb = scalafx_indxl2g(mLoc, desc(MB_), env%blacs%orbitalGrid%myrow, desc(RSRC_), &
+                & env%blacs%orbitalGrid%nrow)
+            if (mGlb >= denseDesc%iAtomStart(iAtom) .and. &
+                & mGlb < denseDesc%iAtomStart(iAtom + 1)) then
+              maskMat(mLoc,:) = sTimesGrndEigVecs(mLoc,:,iSpin)
+            end if
+          end do
+          
+          call pblasfx_pgemm(grndEigVecs(:,:,iSpin), denseDesc%blacsOrbSqr, maskMat, &
```

Comment:
As this is inside a loop over atoms, doesn't this make the routine O(N^3) scaling? (~constant number of matrix vector products * natoms)

Reply:
We can look at this again after the branch is merged.